frontend/src/pages/Graphing/components/SqlEditor.tsx

```diff
@@ -70,6 +71,139 @@ export const DEFAULT_ALERT_SQL = [
+			return `sum(toFloat64OrNull(${columnStr}))`
+		case MetricAggregator.Avg:
+			return `avg(toFloat64OrNull(${columnStr}))`
+		case MetricAggregator.Min:
+			return `min(toFloat64OrNull(${columnStr}))`
+		case MetricAggregator.Max:
+			return `max(toFloat64OrNull(${columnStr}))`
+		case MetricAggregator.Count:
+			return 'count()'
+		case MetricAggregator.CountDistinct:
+			return `uniqExact(${columnStr})`
+		case MetricAggregator.P50:
+			return `quantile(0.50)(toFloat64OrNull(${columnStr}))`
+		case MetricAggregator.P90:
+			return `quantile(0.90)(toFloat64OrNull(${columnStr}))`
+		case MetricAggregator.P95:
+			return `quantile(0.95)(toFloat64OrNull(${columnStr}))`
+		case MetricAggregator.P99:
+			return `quantile(0.99)(toFloat64OrNull(${columnStr}))`
+	}
+}
+
+const productTypeToTable = (productType: ProductType) => {
+	switch (productType) {
+		case ProductType.Sessions:
+			return 'sessions'
+		case ProductType.Logs:
+			return 'logs'
+		case ProductType.Traces:
+			return 'traces'
+		case ProductType.Events:
+			return 'events'
+		case ProductType.Errors:
+			return 'errors'
+		case ProductType.Metrics:
+			return 'metrics'
+	}
+}
+
+const frequencyToName = Object.fromEntries(
+	BUCKET_FREQUENCIES.map((f) => [f.value, f.name]),
+)
+
+const bucketByExpr = (
+	bucketByEnabled: boolean,
+	bucketBySetting: 'Interval' | 'Count',
+	bucketByKey: string,
+	bucketCount: number,
+	bucketInterval: number,
+	startDate: Date,
+	endDate: Date,
+) => {
+	if (!bucketByEnabled) {
+		return undefined
+	}
+
+	switch (bucketBySetting) {
+		case 'Interval':
+			return `$time_interval('${frequencyToName[bucketInterval] ?? '1 hour'}')`
+		case 'Count':
+			if (bucketByKey !== TIMESTAMP_KEY) {
+				return undefined
+			}
+			const startUnix = Math.floor(startDate.getTime() / 1000)
+			const endUnix = Math.floor(endDate.getTime() / 1000)
+			const bucketSeconds = Math.ceil((endUnix - startUnix) / bucketCount)
+			return `fromUnixTimestamp(intDiv(toUnixTimestamp(timestamp) - ${startUnix}, ${bucketSeconds}) * ${bucketSeconds} + ${startUnix})`
+	}
+}
+
+export const convertSettingsToSql = (
+	settings: GraphSettings,
+	startDate: Date,
+	endDate: Date,
+) => {
+	const {
+		productType,
+		query,
+		groupByEnabled,
+		groupByKeys,
+		bucketByEnabled,
+		bucketBySetting,
+		bucketByKey,
+		bucketCount,
+		bucketInterval,
+		expressions,
+	} = settings
+
+	const bucketExpr = bucketByExpr(
+		bucketByEnabled,
+		bucketBySetting,
+		bucketByKey,
+		bucketCount,
+		bucketInterval,
+		startDate,
+		endDate,
+	)
+
+	const groupingExprs = [
+		...(groupByEnabled ? groupByKeys : []),
+		bucketExpr,
+	].filter((e) => e !== undefined)
+
+	const groupByExpr = groupingExprs
+		.map((_, i) => (i + 1).toString())
+		.join(', ')
+
+	const selectExprs = [
+		...groupingExprs,
+		...expressions.map((e) => expressionToSql(e)),
+	]
+
+	let sql = `SELECT ${selectExprs.join(',\n\t')}\nFROM ${productTypeToTable(productType)}`
+	if (query) {
+		sql += `\nWHERE ${query}`
```

Comment:
Only edge case I saw is if there are multiple statements without explicit `AND` and maybe missing some string, but think it should be pretty clear to the user what to fix, so don't think it is blocking since may be a complicated issue

![Screenshot 2025-03-06 at 10 39 57 AM](https://github.com/user-attachments/assets/2a39f051-42ea-4e4e-bb6b-9f4d33471815)

Reply:
good point! I think fixing this is a little tricky, can do later